from distutils.core import setup
import os
import sys

fastqc_files= [
'FastQC/Contaminants/contaminant_list.txt',
'FastQC/Help/1 Introduction/1.1 What is FastQC.html',
'FastQC/Help/2 Basic Operations/2.1 Opening a sequence file.html',
'FastQC/Help/2 Basic Operations/2.2 Evaluating Results.html',
'FastQC/Help/2 Basic Operations/2.3 Saving a Report.html',
'FastQC/Help/3 Analysis Modules/1 Basic Statistics.html',
'FastQC/Help/3 Analysis Modules/10 Overrepresented Sequences.html',
'FastQC/Help/3 Analysis Modules/11 Overrepresented Kmers.html',
'FastQC/Help/3 Analysis Modules/2 Per Base Sequence Quality.html',
'FastQC/Help/3 Analysis Modules/3 Per Sequence Quality Scores.html',
'FastQC/Help/3 Analysis Modules/4 Per Base Sequence Content.html',
'FastQC/Help/3 Analysis Modules/5 Per Base GC Content.html',
'FastQC/Help/3 Analysis Modules/6 Per Sequence GC Content.html',
'FastQC/Help/3 Analysis Modules/7 Per Base N Content.html',
'FastQC/Help/3 Analysis Modules/8 Sequence Length Distribution.html',
'FastQC/Help/3 Analysis Modules/9 Duplicate Sequences.html',
'FastQC/Help/3 Analysis Modules/duplication_levels.png',
'FastQC/Help/3 Analysis Modules/kmer_profiles.png',
'FastQC/Help/3 Analysis Modules/per_base_gc_content.png',
'FastQC/Help/3 Analysis Modules/per_base_n_content.png',
'FastQC/Help/3 Analysis Modules/per_base_quality.png',
'FastQC/Help/3 Analysis Modules/per_base_sequence_content.png',
'FastQC/Help/3 Analysis Modules/per_sequence_gc_content.png',
'FastQC/Help/3 Analysis Modules/per_sequence_quality.png',
'FastQC/Help/3 Analysis Modules/sequence_length_distribution.png',
'FastQC/INSTALL.txt',
'FastQC/LICENSE.txt',
'FastQC/README.txt',
'FastQC/RELEASE_NOTES.txt',
'FastQC/Templates/Icons/error.png',
'FastQC/Templates/Icons/fastqc_icon.png',
'FastQC/Templates/Icons/tick.png',
'FastQC/Templates/Icons/warning.png',
'FastQC/Templates/header_template.html',
'FastQC/fastqc',
'FastQC/fastqc_icon.ico',
'FastQC/jbzip2-0.9.jar',
'FastQC/run_fastqc.bat',
'FastQC/sam-1.32.jar',
'FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisListener.class',
'FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisQueue.class',
'FastQC/uk/ac/babraham/FastQC/Analysis/AnalysisRunner.class',
'FastQC/uk/ac/babraham/FastQC/Analysis/OfflineRunner.class',
'FastQC/uk/ac/babraham/FastQC/Dialogs/AboutDialog$1.class',
'FastQC/uk/ac/babraham/FastQC/Dialogs/AboutDialog.class',
'FastQC/uk/ac/babraham/FastQC/Dialogs/FastQCTitlePanel$SmoothJLabel.class',
'FastQC/uk/ac/babraham/FastQC/Dialogs/FastQCTitlePanel.class',
'FastQC/uk/ac/babraham/FastQC/Dialogs/WelcomePanel.class',
'FastQC/uk/ac/babraham/FastQC/FastQCApplication$1.class',
'FastQC/uk/ac/babraham/FastQC/FastQCApplication.class',
'FastQC/uk/ac/babraham/FastQC/FastQCMenuBar.class',
'FastQC/uk/ac/babraham/FastQC/FileFilters/BAMFileFilter.class',
'FastQC/uk/ac/babraham/FastQC/FileFilters/CasavaFastQFileFilter.class',
'FastQC/uk/ac/babraham/FastQC/FileFilters/FastQFileFilter.class',
'FastQC/uk/ac/babraham/FastQC/FileFilters/GobyFileFilter.class',
'FastQC/uk/ac/babraham/FastQC/FileFilters/MappedBAMFileFilter.class',
'FastQC/uk/ac/babraham/FastQC/FileFilters/SequenceFileFilter.class',
'FastQC/uk/ac/babraham/FastQC/Graphs/BaseGroup.class',
'FastQC/uk/ac/babraham/FastQC/Graphs/LineGraph.class',
'FastQC/uk/ac/babraham/FastQC/Graphs/QualityBoxPlot.class',
'FastQC/uk/ac/babraham/FastQC/Help/HelpDialog.class',
'FastQC/uk/ac/babraham/FastQC/Help/HelpIndexRoot$FileSorter.class',
'FastQC/uk/ac/babraham/FastQC/Help/HelpIndexRoot.class',
'FastQC/uk/ac/babraham/FastQC/Help/HelpPage.class',
'FastQC/uk/ac/babraham/FastQC/Help/HelpPageDisplay$HelpEditor.class',
'FastQC/uk/ac/babraham/FastQC/Help/HelpPageDisplay.class',
'FastQC/uk/ac/babraham/FastQC/Help/HelpSearchPanel.class',
'FastQC/uk/ac/babraham/FastQC/Modules/BasicStats$ResultsTable.class',
'FastQC/uk/ac/babraham/FastQC/Modules/BasicStats.class',
'FastQC/uk/ac/babraham/FastQC/Modules/DuplicationLevel.class',
'FastQC/uk/ac/babraham/FastQC/Modules/GCModel/GCModel.class',
'FastQC/uk/ac/babraham/FastQC/Modules/GCModel/GCModelValue.class',
'FastQC/uk/ac/babraham/FastQC/Modules/KmerContent$Kmer.class',
'FastQC/uk/ac/babraham/FastQC/Modules/KmerContent$ResultsTable.class',
'FastQC/uk/ac/babraham/FastQC/Modules/KmerContent.class',
'FastQC/uk/ac/babraham/FastQC/Modules/NContent.class',
'FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs$OverrepresentedSeq.class',
'FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs$ResultsTable.class',
'FastQC/uk/ac/babraham/FastQC/Modules/OverRepresentedSeqs.class',
'FastQC/uk/ac/babraham/FastQC/Modules/PerBaseGCContent.class',
'FastQC/uk/ac/babraham/FastQC/Modules/PerBaseQualityScores$QualityCount.class',
'FastQC/uk/ac/babraham/FastQC/Modules/PerBaseQualityScores.class',
'FastQC/uk/ac/babraham/FastQC/Modules/PerBaseSequenceContent.class',
'FastQC/uk/ac/babraham/FastQC/Modules/PerSequenceGCContent.class',
'FastQC/uk/ac/babraham/FastQC/Modules/PerSequenceQualityScores.class',
'FastQC/uk/ac/babraham/FastQC/Modules/QCModule.class',
'FastQC/uk/ac/babraham/FastQC/Modules/SequenceLengthDistribution.class',
'FastQC/uk/ac/babraham/FastQC/Report/HTMLReportArchive.class',
'FastQC/uk/ac/babraham/FastQC/Report/stylesheet.txt',
'FastQC/uk/ac/babraham/FastQC/Resources/error.png',
'FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon.png',
'FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon.svg',
'FastQC/uk/ac/babraham/FastQC/Resources/fastqc_icon_100.png',
'FastQC/uk/ac/babraham/FastQC/Resources/tick.png',
'FastQC/uk/ac/babraham/FastQC/Resources/warning.png',
'FastQC/uk/ac/babraham/FastQC/Results/ResultsPanel$ModuleRenderer.class',
'FastQC/uk/ac/babraham/FastQC/Results/ResultsPanel.class',
'FastQC/uk/ac/babraham/FastQC/Sequence/BAMFile.class',
'FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/Contaminant.class',
'FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/ContaminantHit.class',
'FastQC/uk/ac/babraham/FastQC/Sequence/Contaminant/ContaminentFinder.class',
'FastQC/uk/ac/babraham/FastQC/Sequence/FastQFile.class',
'FastQC/uk/ac/babraham/FastQC/Sequence/QualityEncoding/PhredEncoding.class',
'FastQC/uk/ac/babraham/FastQC/Sequence/Sequence.class',
'FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFactory.class',
'FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFile.class',
'FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFileGroup.class',
'FastQC/uk/ac/babraham/FastQC/Sequence/SequenceFormatException.class',
'FastQC/uk/ac/babraham/FastQC/Statistics/NormalDistribution.class',
'FastQC/uk/ac/babraham/FastQC/Statistics/PearsonCorrelation.class',
'FastQC/uk/ac/babraham/FastQC/Utilities/CasavaBasename.class',
'FastQC/uk/ac/babraham/FastQC/Utilities/MultiMemberGZIPInputStream.class',
'FastQC/uk/ac/babraham/FastQC/Utilities/NameFormatException.class']



setup(
    name = 'bsExpress',
    version = '0.5cegx',
    author = 'Dario Beraldi',
    author_email = 'dario.beraldi<at>gmail<dot>com',
    url= '',
    description = "Pipeline to analyze and quality check (ox)Bisulfite libraries",
    classifiers = [
        'Topic :: Scientific/Engineering :: Bio-Informatics',
        'Intended Audience :: Science/Research',
        'License :: OSI Approved :: GNU General Public License (GPL)',
        'Operating System :: POSIX',
        'Programming Language :: Python'],

    requires = [ 'argparse', ],
   
    py_modules = [
        'oxbs_qc.cleanReadNames',
        'oxbs_qc.mpileup2methylation',
        'oxbs_qc.oxbs_qc',
        'oxbs_qc.oxbs_qc_func',
        'oxbs_qc.validate_args'], 
 
    scripts = [
        'scripts/bsExpress',],

    packages= ['oxbs_qc'],
    package_data= {'oxbs_qc': ['MarkDuplicates.jar', 'ShortenFastq.jar', 'CleanReadNames.jar', 'bismark', 'oxbs_report.R', 'trim_galore'] + fastqc_files}
)


